plugin.video.invidious/resources/lib/invidious_plugin.py

```diff
@@ -0,0 +1,338 @@
+            instance_url = self.instance_autodetect()
+        xbmc.log(f'invidous using instance {instance_url}.', xbmc.LOGINFO)
+        self.api_client = invidious_api.InvidiousAPIClient(instance_url)
+        self.disable_dash = \
+            ("true" == xbmcplugin.getSetting(self.addon_handle,
+                                             "disable_dash"))
+    def instance_autodetect(self):
+        xbmc.log('invidious picking instance automatically.', xbmc.LOGINFO)
+
+        response = requests.get(self.INSTANCESURL, timeout=5)
+        data = response.json()
+        for instanceinfo in data:
+            xbmc.log('invidious considering instance ' + str(instanceinfo), xbmc.LOGDEBUG)
+            instancename, instance = instanceinfo
+            if 'https' == instance['type']:
+                instance_url = instance['uri']
+                # Make sure the instance work for us.  This test avoid
+                # those rejecting us with HTTP status 429.
+                api_client = invidious_api.InvidiousAPIClient(instance_url)
+                if api_client.fetch_special_list(self.SPECIAL_LISTS[0]):
+                    return instance_url
+
+        xbmc.log('invidious no working https type instance returned from api.invidious.io.', xbmc.LOGWARNING)
+        # FIXME figure out how to show failing autodetection to the user.
+        dialog = xbmcgui.Dialog()
+        dialog.notification(
+            'No working instance URL found',
+            'No working https type instance returned from api.invidious.io.'
+            "error"
+        )
+        raise ValueError("unable to find working Invidious instance")
+
+    def build_url(self, action, **kwargs):
+        if not action:
+            raise ValueError("you need to specify an action")
+
+        kwargs["action"] = action
+
+        return self.base_url + "?" + urlencode(kwargs)
+
+    def add_directory_item(self, *args, **kwargs):
+        xbmcplugin.addDirectoryItem(self.addon_handle, *args, **kwargs)
+
+    def end_of_directory(self):
+        xbmcplugin.endOfDirectory(self.addon_handle)
+
+    def display_search_results(self, results):
+        # FIXME Add pagination support?
+        for result in results:
+            if result.type not in ['video', 'channel', 'playlist']:
+                raise RuntimeError("unknown result type " + result.type)
+
+            list_item = xbmcgui.ListItem(result.heading)
+            list_item.setArt({
+                "thumb": result.thumbnail_url,
+            })
+
+            # if this is NOT set, the plugin is called with an invalid handle when trying to play this item
+            # seriously, Kodi? come on...
+            # https://forum.kodi.tv/showthread.php?tid=173986&pid=1519987#pid1519987
+            list_item.setProperty("IsPlayable", "true")
+
+            if 'video' == result.type:
+                datestr = datetime.utcfromtimestamp(result.published).date().isoformat()
+
+                list_item.setInfo("video", {
```

Comment:
setInfo is deprecated, please migrate to the new InfoTagVideo method. If you are not familiar thread below may help: https://forum.kodi.tv/showthread.php?tid=369255

Reply:
Is this a requirement for distribution?  I tried earlier to make this transition, after noticing a log entry about deprication, but could not make it work.  My draft is in https://github.com/petterreinholdtsen/kodi-invidious-plugin/tree/obsolete-setInfo, and I would love to make the addon available in the repo to the general public while I figure out or get help on how to drop the setInfo() call.

Reply:
Sadly I can not get the infotagger working either.  When trying to use it, the add-on refuse to install from zip with the message "The dependency on script.module.infotagger version 0.0.7 could not be satisfied.".  Also tried version 0.0.3 as instructed on https://github.com/jurialmunkey/script.module.infotagger without any luck.  My current draft is available from https://github.com/petterreinholdtsen/kodi-invidious-plugin/tree/obsolete-setInfo.

Is there any working recipe available on the web somewhere on how to convert setInfo() calls to the new InfoTagVideo approach?  When I tried I ended up with a object with no set*() methods, only get*() methods.  This make me believe I have completely misunderstood how to do this.

Reply:
> Thanks for the pointer. Is the new API so broken that one have to use a wrapper to get it working? I am reluctant to add a new dependency just for this, but have so far failed to get the transition working without it and will try again using this infotagger extra dependency and see if I have more luck with this approach.

If you don't understand the new API there are topics on the Kodi forums to help; ie. your lack of understanding != broken API. The wrapper is there to offer backward/forward compatibility...  in your case since you are only familiar with setinfo, its a bridge to the new API.

Reply:
> Sadly I can not get the infotagger working either. When trying to use it, the add-on refuse to install from zip with the message "The dependency on script.module.infotagger version 0.0.7 could not be satisfied.". Also tried version 0.0.3 as instructed on https://github.com/jurialmunkey/script.module.infotagger without any luck. My current draft is available from https://github.com/petterreinholdtsen/kodi-invidious-plugin/tree/obsolete-setInfo.
> 
> Is there any working recipe available on the web somewhere on how to convert setInfo() calls to the new InfoTagVideo approach? When I tried I ended up with a object with no set*() methods, only get*() methods. This make me believe I have completely misunderstood how to do this.

The readme attached to the infotagger project has excellent examples.